File: examples/ndvi/ndvi.R
set.seed(777)

#
# Libraries
#
library(sf)
library(sits)

#
# General definitions
#

# Collection
collection_name <- "LANDSAT-OLI-16D"

# Source
collection_source <- "BDC"

# Bands
collection_bands <- c("RED", "NIR08", "CLOUD")

# Start date
start_date <- "2019-05-01"

# End date
end_date <- "2019-07-01"

# ROI file
roi_file <- "data/raw/roi/sinop_roi.shp"

# Output directory
output_dir <- "data/derived/sits-ndvi"

#
# 1. Create directories
#

# Output directory
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)


#
# 2. Load ROI file
#
roi_data <- st_read(roi_file)

#
# 3. Create data cube
#
cube <- sits_cube(
  source     = collection_source,
  collection = collection_name,
  bands      = collection_bands,
  roi        = roi_data,
  start_date = start_date,
  end_date   = end_date,
  progress   = TRUE
)

#
# 4. Generate NDVI
#
cube_ndvi <- sits_apply(
  data       = cube,
  NDVI       = (NIR08 - RED) / (NIR08 + RED),
  output_dir = tempdir(),
  multicores = 4,
  progress   = TRUE
)

#
# 5. Crop results
#
cube_ndvi <- sits_mosaic(
  cube       = cube_ndvi,
  roi        = roi_data,
  output_dir = output_dir,
  multicores = 4
)

#
# 6. View results
#

# Static plot
plot(cube_ndvi, band = "NDVI", date = "2019-05-25")

# Interactive plot
sits_view(cube_ndvi, band = "NDVI", date = "2019-05-25", opacity = 1)

File: examples/ndvi/ndvi.js
//
// Auxiliary functions
//

/**
 * Apply Landsat-8 Scale factors to have Surface Reflectance data available.
 *
 * @param image ee.Image Image to apply the scale factor.
 */
function applyLS8ScaleFactor(image) {
  var opticalBands = image.select('SR_B.').multiply(0.0000275).add(-0.2);
  return image.addBands(opticalBands, null, true);
}

/**
 * Mask cloud pixels using QA_BAND.
 *
 * @param image ee.Image Image to mask clouds
 */
function maskClouds(image) {
  // Select the pixel_qa band
  var qa = image.select('QA_PIXEL');

  // Create a mask to identify cloud and cloud shadow
  var cloudMask = qa.bitwiseAnd(1 << 5).eq(0) // Clouds
                      .and(qa.bitwiseAnd(1 << 3).eq(0)); // Cloud shadows

  // Apply the cloud mask to the image
  return image.updateMask(cloudMask);
}

/**
 * Calculate NDVI.
 *
 * @param image ee.Image Image to calculate NDVI
 *
 * @reference https://developers.google.com/earth-engine/tutorials/tutorial_api_06#mapping-a-function-over-a-collection
 */
function calcNDVI(image) {
  var ndvi = image.normalizedDifference(["SR_B5", "SR_B4"]).rename('NDVI');

  return image.addBands(ndvi);
}

//
// General definitions
//

// Start date
var startDate = "2019-05-01";

// End date
var endDate = "2019-07-01";

// Collection
var collectionName = "LANDSAT/LC08/C02/T1_L2";

// bands
var collectionBands = ["SR_B4", "SR_B5", "QA_PIXEL"];

// ROI file
var roiFile = "users/efelipecarlos/sits-gee/sinop_roi";

//
// 1. Load ROI data
//
var roiData = ee.FeatureCollection(roiFile);

//
// 2. Load collection
//
var imageCollection = ee.ImageCollection(collectionName)
                        .filterBounds(roiData)
                        .filterDate(startDate, endDate)
                        .select(collectionBands);

//
// 3. Apply scale factor
//
var imageCollectionScaled = imageCollection.map(applyLS8ScaleFactor);

//
// 4. Map cloud mask
//
var imageCollectionMasked = imageCollectionScaled.map(maskClouds)

//
// 5. Generate NDVI
//
var imageCollectionWithNDVI = imageCollectionMasked.map(calcNDVI);

//
// 6. Select NDVI
//
var imageCollectionNDVI = imageCollectionWithNDVI.select("NDVI");

//
// 7. Crop result
//
imageCollectionNDVI = imageCollectionNDVI.map(function(image) {
  return image.clip(roiData);
});

//
// 8. View results
//

// Define view region
Map.centerObject(roiData, 10)

// Add classification map (colors from sits R package)
Map.addLayer(imageCollectionNDVI, {min: 0, max: 1, palette: ["red", 'white', 'green']}, "NDVI Image");

//
// 9. Export results
//
Export.image.toDrive({
  image: imageCollectionNDVI,
  description: 'NDVI_Sinop',
  scale: 30,
  region: roiData,
  fileFormat: 'GeoTIFF',
  maxPixels: 1e13,
  crs: "EPSG:3857"
});
